### DISCUSSION TITLE
Why doesn't RegexReplacementConverter support patterns for substitution?

### DISCUSSION THREAD

**Discussion by lw33** [apache/logging-log4j2#2845] [Q&A] [answered]
I think supporting patterns would make it more flexible and useful.

**Comment by vy** [comment]
Could you provide us some example use cases? Do you have a particular
need/problem that you are trying to address?

**Reply by lw33** [reply]
In some cases, my single log entry spans multiple lines, and using log collection tools like Filebeat might truncate such logs. Therefore, I want to rewrite these multi-line logs into single-line logs by replacing the line breaks. If the replace functionality supports the following configuration, I can conveniently achieve this functionality.

` <PatternLayout pattern="%d{yyyy-MM-dd HH:mm:ss} %p [%t] %replace{%m}{\n}{%n%d{yyyy/MM/dd HH:mm:ss} %p [%t]}%n" />`

A single log entry contains multiple lines.
```
2024-08-16 15:52:51 INFO [main] testtesttesttesttesttesttesttesttesttest
testtesttesttesttesttesttesttesttesttest
testtesttesttesttesttesttesttesttesttest
testtesttesttesttesttesttesttesttesttest
testtesttesttesttesttesttesttesttesttest
testtesttesttesttesttesttesttesttesttest
testtesttesttesttesttesttesttesttesttest
testtesttesttesttesttesttesttesttesttest
```

I want them to be output as multiple lines.
```
2024-08-16 15:53:25 INFO [main] testtesttesttesttesttesttesttesttesttest
2024/08/16 15:53:25 INFO [main]testtesttesttesttesttesttesttesttesttest
2024/08/16 15:53:25 INFO [main]testtesttesttesttesttesttesttesttesttest
2024/08/16 15:53:25 INFO [main]testtesttesttesttesttesttesttesttesttest
2024/08/16 15:53:25 INFO [main]testtesttesttesttesttesttesttesttesttest
2024/08/16 15:53:25 INFO [main]testtesttesttesttesttesttesttesttesttest
2024/08/16 15:53:25 INFO [main]testtesttesttesttesttesttesttesttesttest
2024/08/16 15:53:25 INFO [main]testtesttesttesttesttesttesttesttesttest
```

**Comment by vy** [comment, marked as answer]
I strongly discourage you from using Pattern Layout for structured
logging[1] purposes. What if thread name contains a newline character? Or
level? PL will easily fall short on adressing such corner cases, neither it
was intended to do so. You are using the wrong tool. You should rather
employ JSON Template Layout for structured logging. It is not only super
customizable and efficient, it is designed with machine-to-machine
communication (e.g., Filebeat feed) in mind.

[1]
https://logging.staged.apache.org/log4j/2.x/manual/layouts.html#structured-logging

**Reply by lw33** [reply]
Thank you for your advice.
